builder/oracle/bmcs/ssh.go

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2017 Oracle America, Inc.
+// The contents of this file are subject to the Mozilla Public License Version
+// 2.0 (the "License"); you may not use this file except in compliance with the
+// License. If a copy of the MPL was not distributed with this file, You can
+// obtain one at http://mozilla.org/MPL/2.0/
+
+package bmcs
+
+import (
+	"fmt"
+
+	packerssh "github.com/hashicorp/packer/communicator/ssh"
+	"github.com/mitchellh/multistep"
+	"golang.org/x/crypto/ssh"
+)
+
+func commHost(state multistep.StateBag) (string, error) {
+	ipAddress := state.Get("instance_ip").(string)
+	return ipAddress, nil
+}
+
+func sshConfig(state multistep.StateBag) (*ssh.ClientConfig, error) {
+	c := state.Get("config").(*Config)
+	privateKey := state.Get("privateKey").(string)
+
+	signer, err := ssh.ParsePrivateKey([]byte(privateKey))
+	if err != nil {
+		return nil, fmt.Errorf("Error setting up SSH config: %s", err)
+	}
+
+	return &ssh.ClientConfig{
+		User: c.Comm.SSHUsername,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+			ssh.Password(c.Comm.SSHPassword),
+			ssh.KeyboardInteractive(
+				packerssh.PasswordKeyboardInteractive(c.Comm.SSHPassword)),
+		},
+	}, nil
```

Comment:
@rickard-von-essen I'm not quite sure how we would go about doing this in a user-friendly manner.

To add an authorised key to a BMCS instance you have to provide the public key when creating it. This being the case we could support SSH Agent by doing one of the following (as far as I can see):

 1. By instantiating a SSH agent client, calling `.List()`, and adding all registered public 
keys to the instance. This doesn't seem like a particularly good idea.
 2. Requiring SSH agent users set `"ssh_agent_key_comment"` and then filter the output of `.List()` using that string. I'm not sure that this is particularly user friendly either but it seems preferable to 1.
 3. Requiring that (when using using SSH agent) the public key must be baked into the source image.

The Amazon builder, for example, gets around this by either [requiring the key pair to be present in the source AMI][1] or [requiring the user to specify a ("could") key pair name][2] which it then associates with the instance via their API.  

[1]: https://github.com/hashicorp/packer/blob/ef274d398db8670d551a92e70c63341d4484c437/builder/amazon/common/step_key_pair.go#L43
[2]: https://github.com/hashicorp/packer/blob/ef274d398db8670d551a92e70c63341d4484c437/builder/amazon/common/step_key_pair.go#L48